linen_examples/pixelcnn/model_test.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Tests for PixelCNN Modules."""
+
+import pixelcnn
+from flax import linen as nn
+from absl.testing import absltest
+from absl.testing import parameterized
+
+import numpy.testing as onp_testing
+
+from jax import random
+import jax.numpy as np
+from jax.config import config
+config.enable_omnistaging()
+
+
+class ModelTest(absltest.TestCase):
+
+  def setUp(self):
+    super(absltest.TestCase, self).setUp()
+    rng = random.PRNGKey(0)
+    self.x = np.arange(24).reshape(1, 4, 3, 2)
+    self.init_input = np.zeros((1, 4, 3, 2), np.float32)
+    self.rng_dict = {'param': rng, 'weightnorm_params': rng}
+
+
+  def init_module(self, module_fn):
+    return module_fn().init(self.rng_dict, self.init_input)['param']
+
+
+  def apply_module(self, module_fn, params):
+    return module_fn().apply(params, self.x, rngs=self.rng_dict)
+
+
+  def get_weightnorm(self, params):
+    return [params[k] for k in ('direction', 'scale', 'bias')]
+
+
+  def assert_out_close_to_zero(self, out):
+    # Weightnorm should ensure that, at initialization time, the outputs of the
+    # module have mean 0 and variance 1 over the non-feature dimensions.
+    # onp_testing.assert_allclose(np.mean(out, (0, 1, 2)), 0., atol=1e-5)
+    onp_testing.assert_allclose(np.var(out, (0, 1, 2)), 1., atol=1e-5)
+
+
+  def test_conv(self):
+    module_fn = lambda: pixelcnn.Conv(features=4, kernel_size=(3, 2))
+    params = self.init_module(module_fn)
+    out = self.apply_module(module_fn, params)
+    weightnorm_params = params['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(weightnorm_params)
+
+    self.assertEqual(direction.shape, (3, 2, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 2, 2, 4))
+    self.assert_close_to_zero(out)
+
+
+  def test_conv_down(self):
+    module_fn = lambda: pixelcnn.ConvDown(features=4)
+    params = self.init_module(module_fn)
+    out = self.apply_module(module_fn, params)
+    weightnorm_params = params['Conv_0']['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(weightnorm_params)
+
+    self.assertEqual(direction.shape, (2, 3, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 4, 3, 4))
+    self.assert_out_close_to_zero(out)
```

Comment:
Per your other comment about this test failing, you can change the logic here to use `module.init_and_output` that reproduces the behavior we had before.

That being said I'm not sure why this would have behaved differently, it would be a bit slower than taking the output during initialization but shouldn't lead to different outcomes.

Reply:
Ah yes I see. In this code you initialize with a different batch of data than what you predict with. So indeed this assertion should fail in this case. Data dependent init only guarantees that the mean and variance are set to 0 and 1 respectively for the same initial batch that you initialized with.

I think if you replace `setUp` so that `self.x` and `self.init_input` are the same it should work, I think (why is there a separation anyways?). But also probably good to use `module.init_and_output` that reproduces what we originally had.

Reply:
Thanks for the tip! I'm now using `Module.init_with_output`. The code looks nicer but the two tests are still failing. I'm happy to dig deeper into this, but maybe you can take a quick look and see whether you think I'm using it correctly?

Reply:
To clarify, `test_conv_down_right` and `test_conv_transpose_down` are still failing. Example error:

```
>     onp_testing.assert_allclose(np.mean(out, (0, 1, 2)), 0., atol=1e-5)
E     AssertionError: 
E     Not equal to tolerance rtol=1e-07, atol=1e-05
E     
E     Mismatched elements: 4 / 4 (100%)
E     Max absolute difference: 0.13608912
E     Max relative difference: inf
E      x: array([ 0.047518, -0.103525,  0.02776 , -0.136089], dtype=float32)
E      y: array(0.)
```

Reply:
The tests are passing now! There was a bug in the weight normalization.